src/jit/valuenum.cpp

```diff
@@ -8718,6 +8885,409 @@ bool Compiler::fgValueNumberHelperCall(GenTreeCall* call)
     return modHeap;
 }
 
+// Adds the exception sets for the current tree node which is performing a memory indirection operation
+void Compiler::fgValueNumberAddExceptionSetForIndirection(GenTree* tree)
+{
+    // We should have an Unary operator
+    assert(tree->OperIsUnary());
+    GenTree* baseAddr = tree->gtGetOp1();
+    ssize_t  offset   = 0;
+
+    // We evaluate the baseAddr ValueNumber further in order
+    // to obtain a better value to use for the null check exeception.
+    //
+    ValueNumPair baseVNP = baseAddr->gtVNPair;
+    ValueNum     baseLVN = baseVNP.GetLiberal();
+    ValueNum     baseCVN = baseVNP.GetConservative();
+    VNFuncApp    funcLAttr;
+    VNFuncApp    funcCAttr;
+    bool         doConservative = true;
+
+    while (vnStore->GetVNFunc(baseLVN, &funcLAttr) && (funcLAttr.m_func == (VNFunc)GT_ADD) &&
+           (vnStore->TypeOfVN(baseLVN) == TYP_BYREF))
+    {
+        if (fgIsBigOffset(offset))
+        {
+            // Failure: Exit this loop if we have a "big" offset
+            break;
+        }
+
+        if (vnStore->IsVNConstant(funcLAttr.m_args[1]) && varTypeIsIntegral(vnStore->TypeOfVN(funcLAttr.m_args[1])))
+        {
+            offset += vnStore->CoercedConstantValue<ssize_t>(funcLAttr.m_args[1]);
+            baseLVN = funcLAttr.m_args[0];
+        }
+        else if (vnStore->IsVNConstant(funcLAttr.m_args[0]) &&
+                 varTypeIsIntegral(vnStore->TypeOfVN(funcLAttr.m_args[0])))
+        {
+            offset += vnStore->CoercedConstantValue<ssize_t>(funcLAttr.m_args[0]);
+            baseLVN = funcLAttr.m_args[1];
+        }
+        else
+        {
+            break;
+        }
+
+        // We also try to calculate the new 'baseCVN'
+        //
+        if (doConservative)
+        {
+            if (vnStore->GetVNFunc(baseCVN, &funcCAttr) && (funcCAttr.m_func == (VNFunc)GT_ADD) &&
+                (vnStore->TypeOfVN(baseCVN) == TYP_BYREF))
+            {
+                if (vnStore->IsVNConstant(funcCAttr.m_args[1]) &&
+                    varTypeIsIntegral(vnStore->TypeOfVN(funcCAttr.m_args[1])))
+                {
+                    offset += vnStore->CoercedConstantValue<ssize_t>(funcCAttr.m_args[1]);
```

Comment:
Hmm, I don't understand this. Looks like the total offset is extracted from both liberal and conservative VNs?

Reply:
I will double check

Reply:
I redid this area and fixed it. Thanks